contracts/5/SushiswapGoblin.sol

```diff
@@ -0,0 +1,250 @@
+
+    constructor(
+        address _operator,
+        IMasterChef _masterChef,
+        IUniswapV2Router02 _router,
+        uint256 _pid,
+        address _sushi,
+        Strategy _addStrat,
+        Strategy _liqStrat,
+        uint256 _reinvestBountyBps
+    ) public {
+        operator = _operator;
+        weth = _router.WETH();
+        masterChef = _masterChef;
+        router = _router;
+        factory = IUniswapV2Factory(_router.factory());
+        // Get lpToken and fToken from MasterChef pool
+        pid = _pid;
+        (IERC20 _lpToken, , , ) = masterChef.poolInfo(_pid);
+        lpToken = IUniswapV2Pair(address(_lpToken));
+        address token0 = lpToken.token0();
+        address token1 = lpToken.token1();
+        fToken = token0 == weth ? token1 : token0;
+        sushi = _sushi;
+        addStrat = _addStrat;
+        liqStrat = _liqStrat;
+        okStrats[address(addStrat)] = true;
+        okStrats[address(liqStrat)] = true;
+        reinvestBountyBps = _reinvestBountyBps;
+        lpToken.approve(address(_masterChef), uint256(-1)); // 100% trust in the staking pool
+        lpToken.approve(address(router), uint256(-1)); // 100% trust in the router
+        fToken.safeApprove(address(router), uint256(-1)); // 100% trust in the router
+        _sushi.safeApprove(address(router), uint256(-1)); // 100% trust in the router
+    }
+
+    /// @dev Require that the caller must be an EOA account to avoid flash loans.
+    modifier onlyEOA() {
+        require(msg.sender == tx.origin, "not eoa");
+        _;
+    }
+
+    /// @dev Require that the caller must be the operator (the bank).
+    modifier onlyOperator() {
+        require(msg.sender == operator, "not operator");
+        _;
+    }
+
+    /// @dev Return the entitied LP token balance for the given shares.
+    /// @param share The number of shares to be converted to LP balance.
+    function shareToBalance(uint256 share) public view returns (uint256) {
+        if (totalShare == 0) return share; // When there's no share, 1 share = 1 balance.
+        (uint256 totalBalance, ) = masterChef.userInfo(pid, address(this));
+        return share.mul(totalBalance).div(totalShare);
+    }
+
+    /// @dev Return the number of shares to receive if staking the given LP tokens.
+    /// @param balance the number of LP tokens to be converted to shares.
+    function balanceToShare(uint256 balance) public view returns (uint256) {
+        if (totalShare == 0) return balance; // When there's no share, 1 share = 1 balance.
+        (uint256 totalBalance, ) = masterChef.userInfo(pid, address(this));
+        return balance.mul(totalShare).div(totalBalance);
+    }
+
+    /// @dev Re-invest whatever this worker has earned back to staked LP tokens.
+    function reinvest() public onlyEOA nonReentrant {
+        // 1. Withdraw all the rewards.
+        (uint256 totalBalance, ) = masterChef.userInfo(pid, address(this));
+        masterChef.withdraw(pid, totalBalance);
```

Comment:
We can withdraw 0 to withdraw all rewards without touching LP token

Reply:
Fixed